navapbc/platform-test
# Add pull request environments

## Ticket

Resolves https://github.com/navapbc/template-infra/issues/139

## Changes

Add callable reusable github action workflows that:
- creates a temporary environment in a separate workspace
- updates the temporary environment
- destroys the temporary environment and workspace

Add template github action workflows for an app that calls the above reusable workflows to:
- create a temporary dev environment when a PR is opened or reopened
- update the environment when the PR is updated
- destroys the temporary environment when the PR is closed

Add documentation around PR environments

## Context for reviewers

This is a feature that has proven to be quite useful on a number of projects, enabling product managers and designers to see and collaborate with engineers on in progress features to UAT the feature or otherwise give feedback on it.

This feature will also enable things like end-to-end testing on PRs and accessibility checks on PRs by providing a full environment they can access that is agnostic to the underlying web stack.

## Testing

Note: Ignore the anchore scan failures since those aren't relevant to this PR

1. In creating this PR, the workflow should have created a PR environment and posted a comment on the PR with a link to the environment.

   Workflow run that created the PR environment: https://github.com/navapbc/platform-test/actions/runs/9667271724/job/26668699416?pr=113

   Screenshot of PR environment
   <img width="636" alt="image" src="https://github.com/navapbc/platform-test/assets/447859/1a8ef920-e574-4b98-8f1b-5638a08a723f">

2. Make a change to the hello world app to change the content, and see that it updates in the PR environment but not in the original dev environment at https://platform-test-dev.navateam.com/. This change is purely for testing and won't be included in the template-infra PR.

   Workflow run that created the PR environment: https://github.com/navapbc/platform-test/actions/runs/9667491476/job/26669440163?pr=113

   Screenshot of PR environment
   <img width="634" alt="image" src="https://github.com/navapbc/platform-test/assets/447859/e4e1e4f2-5494-4769-a4e5-ab0c1e821870">

3. Close the PR, and see that the environment is destroyed

   

4. Reopen the PR, and see that the environment is recreated
